// blitz/dist/blitz.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var app = require('@blitzjs/core/app');
var config = require('@blitzjs/core/config');
var document = require('@blitzjs/core/document');
var dynamic = require('@blitzjs/core/dynamic');
var head = require('@blitzjs/core/head');
var core = require('@blitzjs/core');
var server = require('@blitzjs/core/server');
var image = require('next/image');
var script = require('next/script');
var stdlib = require('next/stdlib');

function _interopDefault (e) { return e && e.__esModule ? e : { 'default': e }; }

var image__default = /*#__PURE__*/_interopDefault(image);



Object.defineProperty(exports, 'Image', {
  enumerable: true,
  get: function () {
    return image__default['default'];
  }
});
Object.defineProperty(exports, 'Script', {
  enumerable: true,
  get: function () {
    return script.Script;
  }
});
Object.defineProperty(exports, 'AuthenticationError', {
  enumerable: true,
  get: function () {
    return stdlib.AuthenticationError;
  }
});
Object.defineProperty(exports, 'AuthorizationError', {
  enumerable: true,
  get: function () {
    return stdlib.AuthorizationError;
  }
});
Object.defineProperty(exports, 'CSRFTokenMismatchError', {
  enumerable: true,
  get: function () {
    return stdlib.CSRFTokenMismatchError;
  }
});
Object.defineProperty(exports, 'NotFoundError', {
  enumerable: true,
  get: function () {
    return stdlib.NotFoundError;
  }
});
Object.defineProperty(exports, 'PaginationArgumentError', {
  enumerable: true,
  get: function () {
    return stdlib.PaginationArgumentError;
  }
});
Object.defineProperty(exports, 'RedirectError', {
  enumerable: true,
  get: function () {
    return stdlib.RedirectError;
  }
});
Object.keys(app).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return app[k];
    }
  });
});
Object.keys(config).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return config[k];
    }
  });
});
Object.keys(document).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return document[k];
    }
  });
});
Object.keys(dynamic).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return dynamic[k];
    }
  });
});
Object.keys(head).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return head[k];
    }
  });
});
Object.keys(core).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return core[k];
    }
  });
});
Object.keys(server).forEach(function (k) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return server[k];
    }
  });
});
